src/Symfony/Component/DependencyInjection/ServiceLocator.php

```diff
@@ -45,23 +59,56 @@ public function has($id)
     public function get($id)
     {
         if (!isset($this->factories[$id])) {
-            throw new ServiceNotFoundException($id, null, null, array_keys($this->factories));
+            throw new ServiceNotFoundException($id, end($this->loading) ?: null, null, array(), $this->createServiceNotFoundMessage($id));
         }
 
-        if (true === $factory = $this->factories[$id]) {
-            throw new ServiceCircularReferenceException($id, array($id, $id));
+        if (isset($this->loading[$id])) {
+            $ids = array_values($this->loading);
+            $ids = array_slice($this->loading, array_search($id, $ids));
+            $ids[] = $id;
+
+            throw new ServiceCircularReferenceException($id, $ids);
         }
 
-        $this->factories[$id] = true;
+        $this->loading[$id] = $id;
         try {
-            return $factory();
+            return $this->factories[$id]();
         } finally {
-            $this->factories[$id] = $factory;
+            unset($this->loading[$id]);
         }
     }
 
     public function __invoke($id)
     {
         return isset($this->factories[$id]) ? $this->get($id) : null;
     }
+
+    private function createServiceNotFoundMessage($id)
+    {
+        if ($this->loading) {
+            $msg = sprintf('The service "%s" has a dependency on a non-existent service "%s".', end($this->loading), $id);
+            $msg .= sprintf(' This locator only knows about "%s"', implode('", "', array_keys($this->factories)));
+        } else {
+            $class = debug_backtrace(DEBUG_BACKTRACE_PROVIDE_OBJECT, 3);
+            $class = isset($class['object'][2]) ? get_class($class['object'][2]) : null;
+            $externalId = $this->externalId ?: $class;
+
+            $msg = sprintf('Service "%s" not found: ', $id);
+
+            if ($this->container && ($this->container->has($id) || isset($this->container->getRemovedIds()[$id]))) {
+                $msg .= 'even though it exists in the app\'s container, ';
+            }
+            if ($externalId) {
+                $msg .= sprintf('"%s" only knows about pre-listed services "%s".', $externalId, implode('", "', array_keys($this->factories)));
+            } else {
+                $msg .= sprintf('the current locator only knows about "%s".', implode('", "', array_keys($this->factories)));
+            }
+
+            if ($class && is_subclass_of($class, ServiceSubscriberInterface::class)) {
+                $msg .= sprintf(' Did you forget to declare the depencency using "%s::getSubscribedServices()"?', preg_replace('/([^\\\\]++\\\\)++/', '', $class));
```

Comment:
Not to go *too* crazy, but if the actual class that implements `ServiceSubscriberInterface` is a vendor, then this error message doesn't make much sense. The screenshot of the error in the PR is a good example of this - the correct solution for a controller is probably *not* to override `getSubscribedServices()`.

For the controller situations specifically (which probably would work for any case where `getSubscribedServices()` is actually implemented by a vendor class), the better message might be:

> Instead of using the container, try using dependency injection to get the service.

We could go even more bonkers and do something (for `AbstractController`) about type-hinting something on the controller... but probably that's overboard. We just need to say "Hey! Stop using the container! Do what you see recommended everywhere in the docs instead".